harbor-client/src/cashu_client.rs

```diff
@@ -0,0 +1,334 @@
+        let url = self
+            .mint_url
+            .join_paths(&["v1", "melt", "quote", "bolt11"])?;
+        self.http_post(url, &request).await
+    }
+
+    /// Melt Quote Status
+    async fn get_melt_quote_status(
+        &self,
+        quote_id: &str,
+    ) -> Result<MeltQuoteBolt11Response<String>, Error> {
+        let url = self
+            .mint_url
+            .join_paths(&["v1", "melt", "quote", "bolt11", quote_id])?;
+
+        self.http_get(url).await
+    }
+
+    /// Melt [NUT-05]
+    /// [Nut-08] Lightning fee return if outputs defined
+    async fn post_melt(
+        &self,
+        request: MeltBolt11Request<String>,
+    ) -> Result<MeltQuoteBolt11Response<String>, Error> {
+        let url = self.mint_url.join_paths(&["v1", "melt", "bolt11"])?;
+        self.http_post(url, &request).await
+    }
+
+    /// Swap Token [NUT-03]
+    async fn post_swap(&self, swap_request: SwapRequest) -> Result<SwapResponse, Error> {
+        let url = self.mint_url.join_paths(&["v1", "swap"])?;
+        self.http_post(url, &swap_request).await
+    }
+
+    /// Get Mint Info [NUT-06]
+    async fn get_mint_info(&self) -> Result<MintInfo, Error> {
+        let url = self.mint_url.join_paths(&["v1", "info"])?;
+        self.http_get(url).await
+    }
+
+    /// Spendable check [NUT-07]
+    async fn post_check_state(
+        &self,
+        request: CheckStateRequest,
+    ) -> Result<CheckStateResponse, Error> {
+        let url = self.mint_url.join_paths(&["v1", "checkstate"])?;
+        self.http_post(url, &request).await
+    }
+
+    /// Restore request [NUT-13]
+    async fn post_restore(&self, request: RestoreRequest) -> Result<RestoreResponse, Error> {
+        let url = self.mint_url.join_paths(&["v1", "restore"])?;
+        self.http_post(url, &request).await
+    }
+}
+
+pub fn spawn_lightning_payment_thread(
+    mut sender: Sender<CoreUIMsgPacket>,
+    client: Wallet,
+    storage: Arc<dyn DBConnection + Send + Sync>,
+    quote: MeltQuote,
+    msg_id: Uuid,
+    is_transfer: bool,
+) {
+    spawn(async move {
+        match client.melt(&quote.id).await {
+            Ok(outgoing) => {
+                log::info!(
+                    "Payment completed: {}, preimage: {:?}",
+                    quote.id,
+                    outgoing.preimage
+                );
+                let preimage: [u8; 32] = FromHex::from_hex(&outgoing.preimage.unwrap())
+                    .expect("preimage must be valid hex");
+                let params = if is_transfer {
+                    SendSuccessMsg::Transfer
+                } else {
+                    SendSuccessMsg::Lightning { preimage }
+                };
+                sender
+                    .send(CoreUIMsgPacket {
+                        id: Some(msg_id),
+                        msg: CoreUIMsg::SendSuccess(params),
+                    })
+                    .await
+                    .unwrap();
```

Comment:
Could clean up this CoreUIMsgPacket boilerplate by using a similar method like "msg" that's impld on harborcore:

    // Standard core->ui communication with an id
    // Panics if fails to send
    pub async fn msg(&self, id: Uuid, msg: CoreUIMsg) {
        self.tx
            .clone()
            .send(CoreUIMsgPacket { id: Some(id), msg })
            .await
            .expect("Could not communicate with the UI");
    }

(there's probably a way to make a type out of `Sender<CoreUIMsgPacket>,` or impl the msg method on that instead so it could be used in both places?)

not a huge deal but it hides these unwraps so it's easier to read this stuff

Reply:
i'll do this in a follow up, we do this in a bunch of places